src/ng/directive/ngClass.js

```diff
@@ -8,92 +8,93 @@
 
 function classDirective(name, selector) {
   name = 'ngClass' + name;
-  return ['$animate', function($animate) {
+  return ['$animate','$parse', function($animate,$parse) {
     return {
       restrict: 'AC',
-      link: function(scope, element, attr) {
-        var oldVal;
-
-        scope.$watch(attr[name], ngClassWatchAction, true);
-
-        attr.$observe('class', function(value) {
-          ngClassWatchAction(scope.$eval(attr[name]));
+      compile: function(tElement, tAttr) {
+        var classFn = $parse(tAttr[name], function(value) {
+          return arrayClassesL1(value);
         });
+        return function(scope, element, attr) {
+          var oldClasses;
 
 
-        if (name !== 'ngClass') {
-          scope.$watch('$index', function($index, old$index) {
-            /* eslint-disable no-bitwise */
-            var mod = $index & 1;
-            if (mod !== (old$index & 1)) {
-              var classes = arrayClasses(scope.$eval(attr[name]));
-              if (mod === selector) {
-                addClasses(classes);
-              } else {
-                removeClasses(classes);
-              }
-            }
-            /* eslint-enable */
-          });
-        }
+          scope.$watch(classFn, ngClassWatchAction, classArrayEquals);
```

Comment:
I'll check, I was wrong. It takes classArrayEquals, using true, all tests passes as expected, no extra copies performed (I have also check the $watch code).
So it is updated, Thanks.